### PR TITLE
Include serial number in daemon device output when trusted

### PR DESCRIPTION
**Pull request by superm1**
This moves the storing of the serial number into the daemon and
when the calling process is UID 0 includes it in device output

### CODE REVIEW

File: libfwupd/fwupd-device-private.h [hunk 1]
```diff
@@ -15,6 +15,7 @@ G_BEGIN_DECLS
 
 FwupdDevice	*fwupd_device_from_variant		(GVariant	*data);
 GVariant	*fwupd_device_to_variant		(FwupdDevice	*device);
+GVariant	*fwupd_trusted_device_to_variant	(FwupdDevice	*device);
```
Comment (hughsie) [inline libfwupd/fwupd-device-private.h]: The way we've done this in the past is to add a `fwupd_device_to_variant_full()` function, that has both the `FwupdDevice	*device` and a bitfield we can extend in the future without breaking API. We could then pass something like `FWUPD_DEVICE_FLAG_TRUSTED` to get all the properties.

File: libfwupd/fwupd-device.c [hunk 2]
```diff
@@ -133,6 +134,42 @@ fwupd_device_set_summary (FwupdDevice *device, const gchar *summary)
...
+	g_return_val_if_fail (FWUPD_IS_DEVICE (device), NULL);
+	return priv->serial;
+}
+
+/**
+ * fwupd_device_set_serial:
+ * @device: A #FwupdDevice
+ * @serial: the device serial number
+ *
+ * Sets the device summary.
```
Comment (hughsie) [inline libfwupd/fwupd-device.c]: serial

File: libfwupd/fwupd-device.c [hunk 3]
```diff
@@ -924,18 +963,8 @@ fwupd_device_incorporate (FwupdDevice *self, FwupdDevice *donor)
...
- * Creates a GVariant from the device data.
- *
- * Returns: the GVariant, or %NULL for error
- *
- * Since: 1.0.0
- **/
-GVariant *
-fwupd_device_to_variant (FwupdDevice *device)
+static GVariant *
+fwupd_device_to_variant_common (FwupdDevice *device, gboolean trusted)
```
Comment (hughsie) [inline libfwupd/fwupd-device.c]: maybe use the flag here too?

File: src/fu-device.c [hunk 4]
```diff
@@ -1070,24 +1070,7 @@ fu_device_set_serial (FuDevice *self, const gchar *serial)
...
- * Returns: a string value, or %NULL if never set.
- *
- * Since: 1.0.3
- **/
-const gchar *
-fu_device_get_serial (FuDevice *self)
-{
-	g_return_val_if_fail (FU_IS_DEVICE (self), NULL);
-	return fu_device_get_metadata (self, "serial");
+	fwupd_device_set_serial (FWUPD_DEVICE (self), serial);
```
Comment (hughsie) [inline src/fu-device.c]: I think this wants to be a macro in `src/fu-device.h` now, no?

File: src/fu-main.c [hunk 5]
```diff
@@ -175,14 +175,18 @@ fu_main_engine_percentage_changed_cb (FuEngine *engine,
 }
 
 static GVariant *
-fu_main_device_array_to_variant (GPtrArray *devices)
+fu_main_device_array_to_variant (GPtrArray *devices, gboolean trusted)
```
Comment (hughsie) [inline src/fu-main.c]: same about the flags, then we can just pass them on

File: src/fu-main.c [hunk 6]
```diff
@@ -563,6 +567,33 @@ fu_main_device_id_valid (const gchar *device_id, GError **error)
 	return FALSE;
 }
 
+/* stolen from ConsoleKit2 */
+static gboolean
+get_trusted (FuMainPrivate *priv, const char *sender)
```
Comment (hughsie) [inline src/fu-main.c]: fu_main_get_device_flags_for_sender()?

File: src/fu-main.c [hunk 7]
```diff
@@ -563,6 +567,33 @@ fu_main_device_id_valid (const gchar *device_id, GError **error)
...
+{
+	GVariant *value = NULL;
+	uid_t calling_uid;
+	g_autoptr(GError) error_local = NULL;
```
Comment (hughsie) [inline src/fu-main.c]: I'd be much happier handling the error in the caller, even if that's just a g_warning there...

File: src/fu-main.c [hunk 8]
```diff
@@ -563,6 +567,33 @@ fu_main_device_id_valid (const gchar *device_id, GError **error)
...
+					NULL,
+					&error_local);
+	if (value == NULL) {
+		g_warning ("%s", error_local->message);
+		return FALSE;
+	}
+	g_variant_get (value, "(u)", &calling_uid);
+	g_variant_unref (value);
+
+	return (calling_uid == 0);
```
Comment (hughsie) [inline src/fu-main.c]: no brackets pls

**Comment by superm1** [comment]
Adjusted for all those comments.

File: src/fu-main.c [hunk 9]
```diff
@@ -174,15 +174,49 @@ fu_main_engine_percentage_changed_cb (FuEngine *engine,
...
+
+	value = g_dbus_proxy_call_sync (priv->proxy_uid,
+					"GetConnectionUnixUser",
+					g_variant_new ("(s)", sender),
+					G_DBUS_CALL_FLAGS_NONE,
+					2000,
+					NULL,
+					error);
+	if (value == NULL)
+		return FALSE;
```
Comment (hughsie) [inline src/fu-main.c]: i think you probably want a g_prefix_error here to give some context. Either here or in fu_main_device_array_to_variant.

**Comment by hughsie** [comment]
One tiny nit, and assuming it passes CI, LGTM. Thanks.

**Pull request merged**